neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,358 @@
+            var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
+            foreach (PoolItem item in _sortedLowPrioTransactions)
             {
-                tx = item.Transaction;
-                return true;
+                // Expire old free transactions
+                if (item.Transaction.IsLowPriority && item.Timestamp < lowPriorityCutOffTime) continue;
+
+                _unverifiedTransactions.TryAdd(item.Transaction.Hash, item);
+                _unverifiedSortedLowPriorityTransactions.Add(item);
             }
-            else
+            
+            RemoveOverCapacity();
+            
+            // Clear the verified transactions now, since they all must be reverified.
+            _unsortedTransactions.Clear();
+            _sortedHighPrioTransactions.Clear();
+            _sortedLowPrioTransactions.Clear();
+
+            // If we know about headers of future blocks, no point in verifying transactions from the unverified tx pool
+            // until we get caught up.
+            if (block.Index < Blockchain.Singleton.HeaderHeight)
+                return;
+
+            int maxHighPrioTransactionsPerBlock = MaxHighPriorityTxsPerBlock;
+
+            ReverifyHighPriorityTransactions(maxHighPrioTransactionsPerBlock, MaxSecondsToReverifyHighPrioTx, snapshot);
+            ReverifyLowPriorityTransactions(Settings.Default.MaxFreeTransactionsPerBlock, MaxSecondsToReverifyLowPrioTx,
+                snapshot);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ReverifyHighPriorityTransactions(int count, double secondsTimeout, Snapshot snapshot)
+        {
+            DateTime reverifyCutOffTimeStamp = DateTime.UtcNow.AddSeconds(secondsTimeout);
+            int addedCount = 0;
+            
+            foreach (PoolItem item in _unverifiedSortedHighPriorityTransactions.Reverse().Take(count).ToArray())
             {
-                tx = null;
-                return false;
+                // Re-verify the top fee max high priority transactions that can be verified in a block
+                if (item.Transaction.Verify(snapshot, _unsortedTransactions.Select(p => p.Value.Transaction)))
+                {
+                    _unsortedTransactions.TryAdd(item.Transaction.Hash, item);
+                    _sortedHighPrioTransactions.Add(item);
+                    addedCount++;
+                    _unverifiedTransactions.TryRemove(item.Transaction.Hash, out _);
+                    _unverifiedSortedHighPriorityTransactions.Remove(item);                        
+                }
+
+                if (DateTime.UtcNow > reverifyCutOffTimeStamp) break;
             }
+
+            return addedCount;
         }
 
-        public bool TryGetValue(UInt256 hash, out Transaction tx)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ReverifyLowPriorityTransactions(int count, double secondsTimeout, Snapshot snapshot)
```

Comment:
This method and ReverifyHighPriorityTransactions and very similar.

Then, I thought if this Objects could be created as pairs, the logic would become more compact and, I believe, more simple. What do you think?

Reply:
@vneolho well based on the methods exposed but sorted set they have to use different structure because you need to take 20 from low priority and 480 from high priority. If the sorted set exposed iterating from a specific item (which the underlying binary tree supports) it would work, but unfortunately SortedSet designers didn’t expose that ability.

Reply:
A pair can have different sizes. .

I mean, an array with 2 objects, one for high and another for low.
Then, you just pass the index to the function.

Reply:
Ok, in that case, you just meant to remove the duplicated logic. Sure; I'll take a few minutes to adjust to remove the duplicated logic. It will need to pass a few other parameters though, and it can just pass the set and not need to add a Pair.

Reply:
Done.